runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java

```diff
@@ -145,303 +212,407 @@ private void initializeDataTransfer() {
-    putWrittenBytesMetric(writtenBytesList, metric);
 
-    metricCollector.endMeasurement(physicalTaskId, metric);
+    // Put TaskGroup-unit metrics.
+    final boolean available = serBlockSize >= 0;
+    putReadBytesMetric(available, serBlockSize, encodedBlockSize, metric);
+    metricCollector.endMeasurement(taskGroupId, metric);
+    if (logicalTaskIdPutOnHold == null) {
+      taskGroupStateManager.onTaskGroupStateChanged(TaskGroupState.State.COMPLETE, Optional.empty(), Optional.empty());
+    } else {
+      taskGroupStateManager.onTaskGroupStateChanged(TaskGroupState.State.ON_HOLD,
+          Optional.of(logicalTaskIdPutOnHold),
+          Optional.empty());
+    }
+    LOG.info("{} Complete!", taskGroupId);
   }
 
   /**
-   * Pass on the data to the following tasks.
-   * @param task the task to carry on the data.
+   * Processes an OperatorTask.
+   *
+   * @param dataHandler TaskDataHandler of a task to execute.
    */
-  private void launchMetricCollectionBarrierTask(final MetricCollectionBarrierTask task) {
+  private void runTask(final TaskDataHandler dataHandler, final Object dataElement) {
+    final Task task = dataHandler.getTask();
+    final OutputCollectorImpl outputCollector = dataHandler.getOutputCollector();
     final String physicalTaskId = getPhysicalTaskId(task.getId());
-    final Map<String, Object> metric = new HashMap<>();
-    metricCollector.beginMeasurement(physicalTaskId, metric);
-    long accumulatedSerializedBlockSize = 0;
-    long accumulatedEncodedBlockSize = 0;
-    boolean blockSizeAvailable = true;
-
-    final long readStartTime = System.currentTimeMillis();
-    final BlockingQueue<DataUtil.IteratorWithNumBytes> dataQueue = new LinkedBlockingQueue<>();
-    final AtomicInteger sourceParallelism = new AtomicInteger(0);
-    physicalTaskIdToInputReaderMap.get(physicalTaskId).stream().filter(inputReader -> !inputReader.isSideInputReader())
-        .forEach(inputReader -> {
-          sourceParallelism.getAndAdd(inputReader.getSourceParallelism());
-          inputReader.read().forEach(compFuture -> compFuture.thenAccept(dataQueue::add));
-        });
 
-    final List data = new ArrayList<>();
-    for (int srcTaskNum = 0; srcTaskNum < sourceParallelism.get(); srcTaskNum++) {
-      try {
-        final DataUtil.IteratorWithNumBytes availableData = dataQueue.take();
-        availableData.forEachRemaining(data::add);
-        if (blockSizeAvailable) {
-          try {
-            accumulatedSerializedBlockSize += availableData.getNumSerializedBytes();
-            accumulatedEncodedBlockSize += availableData.getNumEncodedBytes();
-          } catch (final DataUtil.IteratorWithNumBytes.NumBytesNotSupportedException e) {
-            blockSizeAvailable = false;
-          }
-        }
-      } catch (final InterruptedException e) {
-        throw new BlockFetchException(e);
+    // Process element-wise depending on the Task type
+    if (task instanceof BoundedSourceTask) {
```

Comment:
If we pre-build these stuffs as `Callable` in `TaskDataHandler`, we can avoid to doing `instanceof`s per every elements.